chain/epoch-manager/src/adapter.rs

```diff
@@ -221,6 +222,42 @@ pub trait EpochManagerAdapter: Send + Sync {
         }
     }
 
+    /// Returns a vector of chunk headers, each of which corresponds to the chunk in the `prev_block`
+    /// This function is important when the block after `prev_block` has different number of chunks
+    /// from `prev_block` in cases of resharding.
+    /// In block production and processing, often we need to get the previous chunks of chunks
+    /// in the current block, this function provides a way to do so while handling sharding changes
+    /// correctly.
+    /// For example, if `prev_block` has two shards 0, 1 and the block after `prev_block` will have
+    /// 4 shards 0, 1, 2, 3, 0 and 1 split from shard 0 and 2 and 3 split from shard 1.
+    /// `get_prev_chunk_headers(epoch_manager, prev_block)` will return
+    /// `[prev_block.chunks()[0], prev_block.chunks()[0], prev_block.chunks()[1], prev_block.chunks()[1]]`
+    fn get_prev_chunk_headers(&self, prev_block: &Block) -> Result<Vec<ShardChunkHeader>, Error> {
+        let epoch_id = self.get_epoch_id_from_prev_block(prev_block.hash())?;
+        let shard_ids = self.shard_ids(&epoch_id)?;
+
+        let prev_shard_ids = self.get_prev_shard_ids(prev_block.hash(), shard_ids)?;
```

Comment:
optional: Maybe [this TODO](https://github.com/near/nearcore/blob/2e78a6f0204304e20e02f1f1b04a12acc2085133/chain/epoch-manager/src/adapter.rs#L169) could be addressed while we are here.